### PR TITLE
knot-resolver on darwin: patch library loading, run more tests

### PR DESCRIPTION
**Pull request by vcunat**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] N/A Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] N/A Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
I'm curious if OfBorg can confirm success on aarch64-darwin, too.  (I don't have access to a compatible machine.)

File: pkgs/servers/dns/knot-resolver/default.nix [hunk 1]
```diff
@@ -26,6 +27,16 @@ unwrapped = stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ];
 
+  patches = [
+    (fetchpatch {
+      name = "fix-config-tests-on-darwin.patch";
+      url = "https://gitlab.nic.cz/knot/knot-resolver/-/commit/48ad9d436cf80f58c107774c313a561d852148a0.diff";
+      sha256 = "CEX1XkeYLUSe31xUhNdMRMl1VUXtKFCs5noNJaqL5x0=";
+    })
+  ] ++ lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
```
Comment (SuperSandro2000) [inline pkgs/servers/dns/knot-resolver/default.nix:36]: If we keep this, please make it unconditional

File: pkgs/servers/dns/knot-resolver/default.nix [hunk 2]
```diff
@@ -26,6 +27,16 @@ unwrapped = stdenv.mkDerivation rec {
...
+    ./tests1.patch
+  ];
+
   # Path fixups for the NixOS service.
   postPatch = ''
     patch meson.build <<EOF
```
Comment (SuperSandro2000) [inline pkgs/servers/dns/knot-resolver/default.nix:42]:
```suggestion
  postPatch = ''
    # Path fixups for the NixOS service.
    patch meson.build <<EOF
```

Side note: I am not really happy that we inline patch files.

File: pkgs/servers/dns/knot-resolver/default.nix [hunk 3]
```diff
@@ -39,6 +50,13 @@ unwrapped = stdenv.mkDerivation rec {
     # ExecStart can't be overwritten in overrides.
     # We need that to use wrapped executable and correct config file.
     sed '/^ExecStart=/d' -i systemd/kresd@.service.in
+
+    # On x86_64-darwin loading by soname fails to find the libs, surprisingly.
+    # Even though they should already be loaded and they're in RPATH, too.
+    for f in daemon/lua/{kres,zonefile}.lua; do
+      substituteInPlace "$f" \
+        --replace "ffi.load("    "ffi.load('${lib.getLib knot-dns}/lib/' .. "
```
Comment (SuperSandro2000) [inline pkgs/servers/dns/knot-resolver/default.nix:58]:
```suggestion
        --replace "ffi.load(" "ffi.load('${lib.getLib knot-dns}/lib/' .. "
```

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix:36]: Yes, that was the plan, after these OfBorg experiments end.

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 3 above)
Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix:58]: I can go back there, as I wasn't sure.  The intention was to improve readability, as the line has lots of punctuation.

Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix]: OK, done.

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix]: Both `fetchpatch`-ed unconditionally now.

**Comment by vcunat** [comment]
aarch64-darwin success on the last functional change was confirmed by https://logs.nix.ci/?key=nixos/nixpkgs.188194&attempt_id=c455f87f-a262-4bb1-af27-9ab0d87ac4ef

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 2 above)
Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix:45]: I don't feel strongly about that, for such trivial patches.  Fortunately there's no need to splice nix values in here, contrary to the `ffi.load` below.

**Pull request merged**